File: 2025/go/day-1/part-1.go
package day1

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	circleSize    = 100
	startPosition = 50
)

func Day1Part1() {
	startTs := time.Now()
	file, err := os.Open("../inputs/day-1/part-1/input.txt")
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	position := startPosition
	crossingsAtZero := 0

	for scanner.Scan() {
		direction, distance, err := parseLine(strings.TrimSpace(scanner.Text()))
		if err != nil {
			fmt.Printf("Error parsing line: %v\n", err)
			os.Exit(1)
		}

		switch direction {
		case "L":
			position -= distance
		case "R":
			position += distance
		default:
			fmt.Printf("Error parsing line: %v\n", err)
			os.Exit(1)
		}

		position = wrapPosition(position)

		if position == 0 {
			crossingsAtZero++
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		os.Exit(1)
	}

	endTs := time.Now()
	elapsed := endTs.Sub(startTs)
	fmt.Printf("Solution: %d [%d(μs)]\n", crossingsAtZero, elapsed.Microseconds())
}

func wrapPosition(position int) (wrappedPosition int) {
	return ((position % circleSize) + circleSize) % circleSize
}

func parseLine(line string) (direction string, distance int, err error) {
	if len(line) < 2 {
		return "", 0, fmt.Errorf("invalid line: %s", line)
	}

	distance, err = strconv.Atoi(line[1:])
	if err != nil {
		return "", 0, err
	}

	return line[:1], distance, nil
}

File: 2025/ts/src/day-1/part-1.ts
import * as fs from "fs";
import * as path from "path";
import * as readline from "readline";

const PATH = path.join(__dirname, "../../..", "inputs/day-1/part-1/input.txt");
const CIRCLE_SIZE = 100;
const START_POSITION = 50;

function parseLine(line: string): [string, number] {
  if (line.length < 2) {
    return ["", 0];
  }

  return [line[0], parseInt(line.slice(1), 10)];
}

function wrapPosition(position: number): number {
  return ((position % CIRCLE_SIZE) + CIRCLE_SIZE) % CIRCLE_SIZE;
}

export async function part1(): Promise<void> {
  const startTs = performance.now()
  const fileStream = fs.createReadStream(PATH);
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  try {
    let position = START_POSITION;
    let crossingsAtZero = 0;

    for await (const line of rl) {
      const [direction, distance] = parseLine(line.trim());

      if (direction === "L") {
        position -= distance;
      } else if (direction === "R") {
        position += distance;
      }

      position = wrapPosition(position);

      if (position === 0) {
        crossingsAtZero++;
      }
    }

    const endTs = performance.now()
    const elapsed = (endTs - startTs) * 1000;
    console.log(`Solution: ${crossingsAtZero} [${elapsed.toFixed(0)} (μs)]`);
  } finally {
    rl.close();
  }
}
